File: framework/languages/java/e2e/RegistryMessaging/Client/src/main/java/systems/zlink/e2e/registrymessaging/client/Support/ScenarioAssert.java
package systems.zlink.e2e.registrymessaging.client.Support;

import java.util.Arrays;
import java.util.Set;
import systems.zlink.e2e.registrymessaging.shared.Contracts;
import systems.zlink.httpclient.ZLinkHttpClient;

public final class ScenarioAssert {
    private ScenarioAssert() {
    }

    public static void that(boolean condition, String message) {
        if (!condition) {
            throw new IllegalStateException(message);
        }
    }

    public static void expectFailure(Runnable action) {
        try {
            action.run();
        } catch (RuntimeException error) {
            return;
        }
        throw new IllegalStateException("operation unexpectedly succeeded");
    }

    public static String[] evidence(ZLinkHttpClient role) {
        return role.get("/evidence").submit(String[].class).toCompletableFuture().join().body();
    }

    public static String[] waitEvidence(ZLinkHttpClient role, String contains) {
        return role.post("/evidence/wait")
            .body(new Contracts.EvidenceWaitReq(contains))
            .submit(String[].class).toCompletableFuture().join().body();
    }

    public static Contracts.ProfileRes requestProfileEventually(
        ZLinkHttpClient role,
        String marker) {
        long deadline = System.nanoTime() + java.util.concurrent.TimeUnit.SECONDS.toNanos(15);
        RuntimeException lastError = null;
        while (System.nanoTime() < deadline) {
            try {
                return role.post("/profile/request")
                    .body(new Contracts.ProfileReq(marker))
                    .submit(Contracts.ProfileRes.class).toCompletableFuture().join().body();
            } catch (RuntimeException error) {
                lastError = error;
                sleep(100);
            }
        }
        throw new IllegalStateException("timed out waiting for profile request " + marker, lastError);
    }

    public static Set<String> requestUntilProvidersSeen(
        ZLinkHttpClient role,
        String markerPrefix,
        Set<String> expectedProviders) {
        long deadline = System.nanoTime() + java.util.concurrent.TimeUnit.SECONDS.toNanos(30);
        Set<String> providers = new java.util.HashSet<>();
        int index = 0;
        while (System.nanoTime() < deadline && !providers.containsAll(expectedProviders)) {
            Contracts.ProfileRes reply = requestProfileEventually(role, markerPrefix + "-" + index);
            providers.add(reply.providerRid());
            index++;
            sleep(100);
        }
        return providers;
    }

    public static String[] waitAnyEvidence(
        ZLinkHttpClient first,
        ZLinkHttpClient second,
        String contains) {
        long deadline = System.nanoTime() + java.util.concurrent.TimeUnit.SECONDS.toNanos(30);
        String[] combined = concat(evidence(first), evidence(second));
        while (System.nanoTime() < deadline) {
            combined = concat(evidence(first), evidence(second));
            if (Arrays.stream(combined).anyMatch(line -> line.contains(contains))) {
                return combined;
            }
            sleep(100);
        }
        throw new IllegalStateException("timed out waiting for evidence containing " + contains);
    }

    public static String[] concat(String[] first, String[] second) {
        return java.util.stream.Stream.concat(Arrays.stream(first), Arrays.stream(second))
            .toArray(String[]::new);
    }

    private static void sleep(long milliseconds) {
        try {
            Thread.sleep(milliseconds);
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("interrupted while waiting for evidence", error);
        }
    }
}

File: framework/languages/java/e2e-kotlin/RegistryMessaging/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/client/Support/ScenarioAssert.kt
package Support

import systems.zlink.e2e.kotlin.registrymessaging.client.Support
import java.time.Duration
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileRes

object ScenarioAssert {
    fun that(condition: Boolean, message: String) {
        if (!condition) {
            throw IllegalStateException(message)
        }
    }

    fun countNewEvidence(after: Collection<String>, before: Collection<String>, prefix: String, marker: String): Int {
        val beforeCount = before.count { it.contains(prefix) && it.contains(marker) }
        val afterCount = after.count { it.contains(prefix) && it.contains(marker) }
        return afterCount - beforeCount
    }

    fun requestProfileEventually(role: HttpJson, marker: String): ProfileRes {
        return retryUntil("profile request $marker") {
            role.post("/profile/request", ProfileReq(marker))
        }
    }

    fun <T> retryUntil(description: String, action: () -> T): T {
        val deadline = System.nanoTime() + Duration.ofSeconds(15).toNanos()
        var last: RuntimeException? = null
        while (System.nanoTime() < deadline) {
            try {
                return action()
            } catch (error: RuntimeException) {
                last = error
                Thread.sleep(100)
            }
        }
        throw IllegalStateException("Timed out waiting for $description.", last)
    }

    fun requestUntilProvidersSeen(
        role: HttpJson,
        markerPrefix: String,
        expectedProviders: Set<String>,
    ): Set<String> {
        val deadline = System.nanoTime() + Duration.ofSeconds(30).toNanos()
        val providers = mutableSetOf<String>()
        var index = 0
        while (System.nanoTime() < deadline && !providers.containsAll(expectedProviders)) {
            providers += requestProfileEventually(role, "$markerPrefix-$index").providerRid
            index++
            Thread.sleep(100)
        }
        return providers
    }
}
